# HG changeset patch
# User Rik <rik@octave.org>
# Date 1513901422 28800
#      Thu Dec 21 16:10:22 2017 -0800
# Node ID 9a9b2e0596f5ab013b5180177d4d010934f58c7b
# Parent  138c2380521c1ecc344c6800b886cb56e472489d
Don't error with "%c" arg to whos_line_format (bug #52714).

* variables.cc (parse_whos_line_format): Detect special exception case
where modifier is 'c' and command is unrecognized.  Revert modifier
to default of 'r' and change command to 'c'.

diff --git a/libinterp/corefcn/variables.cc b/libinterp/corefcn/variables.cc
--- a/libinterp/corefcn/variables.cc
+++ b/libinterp/corefcn/variables.cc
@@ -1511,16 +1511,24 @@ public:
             else
               items = sscanf (cmd.c_str (), "%c%c%c:%d:%d:%d;",
                               &garbage, &param.modifier, &param.command,
                               &a, &b, &balance) - 1;
 
             if (items < 2)
               error ("whos_line_format: parameter structure without command in whos_line_format");
 
+            // Exception case of bare class command 'c' without modifier 'l/r'
+            if (param.modifier == 'c'
+                && param_string.find (param.command) == std::string::npos)
+              {
+                param.modifier = 'r';
+                param.command = 'c';
+              }
+
             // Insert data into parameter
             param.first_parameter_length = 0;
             pos = param_string.find (param.command);
             if (pos == std::string::npos)
               error ("whos_line_format: '%c' is not a command", param.command);
 
             param.parameter_length = param_length(pos);
             param.text = param_names(pos);
